app/javascript/components/misc/CodeMirror/a11yTabBinding.ts

```diff
@@ -0,0 +1,144 @@
+    clip: 'rect(0, 0, 0, 0)',
+    whiteSpace: 'nowrap',
+    borderWidth: '0',
+  },
+})
+
+// The typing for this seems to be wrong in CodeMirror.
+// Changes doesn't have `inserted` defined on it
+function tabPressed(changes: any): boolean {
+  // @ts-ignore */
+  return changes.inserted.some((insert) => insert.text == '\t')
+}
+
+// This cycles though the desired value every time tab is pressed.
+// This is useful for ensuring that screen-readers don't just
+// think the text is the same as it previously was and ignore it.
+//
+// Currently this cycles whenever tab is pressed, whether or not we announce
+// it via the aria-live region (ie whether the box is blank or not at the time)
+// We may want to change this but I've not yet worked out how.
+//
+// Announcements only seem to happen once per prompt however many
+// times it is cycled into the box (I presume that's a non-spamming feature)
+// so for now I have 5 different prompts which means this is announced a maximum
+// of 5 times. There's also a chance this is just down to something weird
+// in VoiceOver (which seems to be quite random to me) so this might be entirelty
+// unncessary and we can just switch between twodifferent ones as per a previous
+// commit in the PR that adds this.
+const prompts = [
+  'Press Escape then Tab to exit the editor',
+  'To exit the editor, press Escape then tab',
+  'Press Escape followed by tab to exit the editor',
+  'If you want to exit the editor, press Escape then Tab',
+  'In order to exit the editor, press Escape then Tab',
+]
+const a11yTabBindingState = StateField.define<string>({
+  create: () => '',
+  update(value, tr) {
+    if (!tabPressed(tr.changes)) {
+      return value
+    }
+    let found = false
+    const newValue = prompts.find((prompt) => {
+      if (found) {
+        return value
+      }
+
+      if (value == prompt) {
+        found = true
+      }
+    })
+    value = newValue || prompts[0]
+    return value
+  },
+})
+
+// This creates a CodeMirror Panel, adds its classes and sets it
+// to be an assertive ariaLive area. It is assertive as it only
+// changes in direct response to a user pressing the tab key.
+function createA11yTabBindingPanel(view: EditorView): Panel {
+  let dom = document.createElement('div')
+  dom.textContent = ''
+  dom.className = 'cm-a11y-panel'
+
+  // @ts-ignore */
+  dom.ariaLive = 'assertive'
+  // @ts-ignore */
+  dom.ariaAtomic = 'true'
+
+  return {
+    top: true,
+    dom,
+    update(update) {
+      // Only announce things if the user has pressed a tab
+      if (!update.docChanged || !tabPressed(update.changes)) {
+        return
+      }
+
+      // Only announce things if the aria-live area has been reset
+      if (dom.textContent != '') {
+        return
+      }
+
+      // Set the text content, which announces via the aria-live functionality.
+      dom.textContent = view.state.field(a11yTabBindingState)
+
+      // Reset the text content five seconds later.
```

Comment:
I'm assuming you do this because you need to re-announce it the next time? The 5second rules might seem safe, but I can configure my screen reader to s p e a k v  e  r  y  s    l    o    w    e   l   y.

I suspect that what you want to do can *actually* be achieved by changed the `id` of this element when you want to start announcing. I _think_ a new ID will act as if it's a new element. But we should try this in VoiceOver, NVDA and JAWS.

Reply:
The 5 second thing more because if someone does presses `tab tab tab`, it would restart each time (because of assertive: true)

> I suspect that what you want to do can actually be achieved by changed the id of this element when you want to start announcing. I think a new ID will act as if it's a new element. But we should try this in VoiceOver, NVDA and JAWS.

This might solve lots. I'll try this :)